code/modules/modular_computers/file_system/programs/messenger/messenger_program.dm

```diff
@@ -0,0 +1,663 @@
+		data["sending_virus"] = sending_virus
+	return data
+
+//////////////////////
+// MESSAGE HANDLING //
+//////////////////////
+
+/// Brings up the quick reply prompt to send a message.
+/datum/computer_file/program/messenger/proc/quick_reply_prompt(mob/living/user, datum/pda_chat/chat)
+	if(!istype(chat))
+		return
+	var/datum/computer_file/program/messenger/target = chat.recipient?.resolve()
+	if(!istype(target) || !istype(target.computer))
+		to_chat(user, span_notice("ERROR: Recipient no longer exists."))
+		chat.recipient = null
+		chat.can_reply = FALSE
+		return
+	var/target_name = target.computer.saved_identification
+	var/input_message = tgui_input_text(user, "Enter [mime_mode ? "emojis":"a message"]", "NT Messaging[target_name ? " ([target_name])" : ""]", encode = FALSE)
+	send_message(user, input_message, list(chat))
+
+/// Helper proc that sends a message to everyone
+/datum/computer_file/program/messenger/proc/send_message_to_all(mob/living/user, message)
+	var/list/datum/pda_chat/chats = list()
+	var/list/msgr_targets = list()
+
+	for(var/mc in get_messengers())
+		msgr_targets += mc
+
+	for(var/chatref in saved_chats)
+		var/datum/pda_chat/chat = saved_chats[chatref]
+		if(chat.recipient?.reference in msgr_targets) // if its in msgr_targets, it's valid
+			msgr_targets -= chat.recipient.reference
+			chats += chat
+
+	for(var/missing_msgr in msgr_targets)
+		var/datum/pda_chat/new_chat = create_chat(missing_msgr)
+		chats += new_chat
+
+	if(send_message(user, message, chats, everyone = TRUE))
+		COOLDOWN_START(src, last_text_everyone, 2 MINUTES)
+
+/// Creates a chat and adds it to saved_chats. Supports fake users. Returns the newly created chat.
+/datum/computer_file/program/messenger/proc/create_chat(recp_ref, name, job)
+	var/datum/computer_file/program/messenger/recipient = null
+
+	if(isnull(name) && isnull(job))
+		if(!(recp_ref in GLOB.TabletMessengers))
+			CRASH("tried to create a chat with a messenger that isn't registered")
+		recipient = GLOB.TabletMessengers[recp_ref]
+
+	var/datum/pda_chat/new_chat = new(recipient)
+
+	// this is a chat with a "fake user" (automated or forged message)
+	if(!istype(recipient))
+		new_chat.cached_name = name
+		new_chat.cached_job = job
+		new_chat.can_reply = FALSE
+
+	saved_chats[REF(new_chat)] = new_chat
+
+	return new_chat
+
+/// Gets the chat by the recipient, either by their name or messenger ref
+/datum/computer_file/program/messenger/proc/find_chat_by_recp(recipient, fake_user = FALSE)
+	for(var/chat_ref in saved_chats)
+		var/datum/pda_chat/chat = saved_chats[chat_ref]
+		if(fake_user && chat.cached_name == recipient)
+			return chat
+		else if(chat.recipient?.reference == recipient)
+			return chat
+	return null
+
+/// Returns a message input, sanitized and checked against the filter
+/datum/computer_file/program/messenger/proc/sanitize_pda_msg(message, mob/sender)
+	// check message against filter
+	if(!check_pda_msg_against_filter(message, sender))
```

Comment:
You are calling this before it is sanitized, but in this proc you are sending the full message to Admins. Isn't this unsafe?

Reply:
that would be a huge FUCK moment, thanks for calling this out